In this notebook the structure of the CORD-19 metadata.csv file is explored to the end of arriving at an appropriate database structure.

# TOC

1. [Imports](#Imports)
1. [metadata.csv](#metadata_csv)
    1. [Columns](#Columns)
       1. [Single valued and unique](#Single_valued_and_unique)
       1. [Single valued](#Single_valued)
1. [Setting up a schema](#Setting_up_a_schema)
    1. [Main table](#Main_table)
    1. [Multivalued field tables](#Multivalued_field_tables)
       1. [pdf_json_files](#pdf_json_files)
       1. [pmc_json_files](#pmc_json_files)
       1. [sha](#sha)
       1. [authors](#authors)
    1. [Quick test](#Quick_test)

<a id='Imports'></a>
# Imports

In [1]:
import csv
import pandas as pd
from pprint import pprint as pp

In [2]:
DataDirectory = './cord-19_2020-06-01/'

<a id='metadata_csv'></a>
# metadata.csv

In [3]:
metadataFile = open(DataDirectory+'metadata.csv', 'r')
csvReader = csv.reader(metadataFile, delimiter=',', quotechar='"')

Header = next(csvReader)
pp(Header)

['cord_uid',
 'sha',
 'source_x',
 'title',
 'doi',
 'pmcid',
 'pubmed_id',
 'license',
 'abstract',
 'publish_time',
 'authors',
 'journal',
 'mag_id',
 'who_covidence_id',
 'arxiv_id',
 'pdf_json_files',
 'pmc_json_files',
 'url',
 's2_id']


<a id='Columns'></a>
## Columns

Starting from the readme provided with the data (metadata.readme) we can outline what each columns should cover and, importantly for databasing it, whether it is single value of potentially multi-valued.

'cord_uid' - This is a persistent identifier for the article within the CORD-19 ecosystem. __single valued__, __unique__.

'sha' - The hash of the article's PDF. There may be multiple PDFs associated to one article (supporting materials, ?preprints?). __multi-valued__.

'source_x' - The source of the article data. Sources covered include CZI (Chan-Zuckerberg Initiative), PMC (pubmed central), bioRxiv and medRxiv. Confirm sources from data. Should be __single valued__.

'title' - The article's title. __single valued__.

'doi' - The article's doi. __single valued__.

'pmcid' - The article's pubmed central id, if it has one. __single valued__, __unique__.


'pubmed_id' - The article's pubmed id. For distinction between this and the previous, see [here](https://publicaccess.nih.gov/include-pmcid-citations.htm#Difference). __single valued__, __unique__.

'license' - The license under which the data is being shared. Should be __single valued__.

'abstract' - The article's abstract. __single valued__.

'publish_time' - The date the article was published. __single valued__.

'authors' - The article's authors. Coverage questionable. __multi-valued__.

'journal' - The journal in which the article was published. If it was published in a journal. What happens with preprints? __single valued__.

'mag_id' - The article's microsoft academic graph id. __single valued__, __unique__.

'who_covidence_id' - The article's WHO #Covidence id. Should only be populated for CZI source articles. __single valued__, __unique__.

'arxiv_id' - The article's arXiv id. __single valued__, __unique__.

'pdf_json_files' - Relative path of the json(s) parsed from pdf file(s). __multi-valued__, __unique__.

'pmc_json_files' - Relative path of the json(s) parsed from pmc file(s). __multi-valued__, __unique__.

'url' - The article's URL. Should be __single valued__, __unique__.

's2_id' - The article's Semantic Scholar id. Can be resolved at https://api.semanticscholar.org/. Should be __single valued__, __unique__.


To explore the columns, it is probably easier to work with the csv via pandas.

In [4]:
metadataFrame = pd.read_csv(DataDirectory+'metadata.csv', low_memory=False)
metadataFrame

cord_uid                                                sha  \
0       ug7v899j           d1aafb70c066a2068b02786f8929fd9c900897fb   
1       02tnwd4m           6b0567729c2143a66d737eb0a2f63f2dce2e5a7d   
2       ejv2xln0           06ced00a5fc04215949aa72528f2eeaae1d58927   
3       2b73a28n           348055649b6b8cf2b9a376498df9bf41f7123605   
4       9785vg6d           5f48792a5fa08bed9f56016f4981ae2ca6031b32   
5       zjufx4fo           b2897e1277f56641193a6db73825f707eed3e4c9   
6       ymceytj3           e3d0d482ebd9a8ba81c254cc433f314142e72174   
7       wzj2glte           00b1d99e70f779eb4ede50059db469c65e8c1469   
8       2sfqsfm1           cf584e00f637cbd8f1bb35f3f09f5ed07b71aeb0   
9       i0zym7iq           dde02f11923815e6a16a31dd6298c46b109c5dfa   
10      5yhe786e           3bb07ea10432f7738413dff9816809cc90f03f99   
11      8zchiykl           5806726a24dc91de3954001effbdffd7a82d54e2   
12      xqhn0vbp           1e1286db212100993d03cc22374b624f7caee956   
13      gi6uaa83           8ae137c8da1607b3a8e4c946c07ca8bda67f88ac   
14      0m32ecnu           23bc55d6f63fab18b02004483888db2b6a0bfa48   
15      le0ogx1s                                                NaN   
16      8qnrcgnk           faaf1022ccfe93b032c5608097a53543ba24aedb   
17      3oxzzxnd           7ff45096210eeb392d51f646f5c7fe011079aaf3   
18      fy4w7xz8           0104f6ceccf92ae8567a0102f89cbb976969a774   
19      jg13scgo           5b44feca5d6ffaaeb66501fa84cc6dd44d06660a   
20      0qaoam29           5b68a553a7cbbea13472721cd1ad617d42b40c26   
21      qj4dh6rg           3ed670f60a7be2e3e2a991ea8af1fdd5fa5e2b2c   
22      1wswi7us           59ffca26c9d1f881e4e37e7c5af626d72d858dc4   
23      yy96yeu9           c63c4d58d170136b8d3b5a66424b5ac3f73a92d9   
24      5o38ihe0           dab2d95da5307c840eb4a1e00f0186022541d63d   
25      1ul8owic           75ebad1e8641e5fe2a51214f4d97742ab59a8dae   
26      5s6acr7m           6c0dee643345c93436cfa8d58270aa887ad92e3d   
27      5tkvsudh           9d4e3e8eb092d5ed282d0aa4aadcaa8b7165b5e9   
28      6lvn10f4           14e0cac6e86d62859e6c9f1351ab67466d89e5b3   
29      tvxpckxo           d09b79026117ec9faebba46a8d13aa9b23ec751e   
...          ...                                                ...   
140502  22s9umrj           cdbdc0b6176729a11ba5e08c708408592d189657   
140503  icowa7wn           d27842f2883b8884d975fe744e1157bb727d6f9b   
140504  kon3ooo3           fb0bce3654938b8c93e5b324d2a23e90f95bb15b   
140505  ccdj7137           8d1498b4b95b0bf8f6d9577764fffcecd1bae360   
140506  zl91u567           738d255294896f43d6ac7546a891900714ef1d51   
140507  47okncdw           2918a43791587306be37d4e23f804c1efdbdd8fa   
140508  4bheqtk5           21974784c3eb198439eabbd88509396d8ad2c156   
140509  bary8svr           1b759c9a06fa77009c2271b444b2236874e99cbc   
140510  oo5x3rrz           f7c6f6eda0e0bfa77134c4fdbd34ce45a9e9c054   
140511  jhiap68j           621832d547930a8c23859fa443c2c9c38ff07e42   
140512  7kiahjyy           b466c5c6dc539542eba79097001d04e2b3fca72e   
140513  uau5ygys           ff8610a4d772b4ec5497589889f8c8bf8179dd7b   
140514  r41ilao6           6bde3798df501fc40dd89e2cff0a1a8fd2b9dd47   
140515  ke5hxd8o                                                NaN   
140516  9aacmywu                                                NaN   
140517  gubekyum           704fd6c6608f5b400f005042357180f8a4fd029d   
140518  h9m4iexp           556e48d9685a2e212757cc015b35c203ecf31487   
140519  p83we2ye           ff09007b2973bf3dc9a0435228a1b220587763da   
140520  cekegxop           3711609ee9f51aed04b50dc095784795c7e0202d   
140521  hp11at50                                                NaN   
140522  j9nllb4f           ab98d1b125aa0704e63adef426b27abd32e935f0   
140523  qgy3xk73           633e89365f800d4e5c888ad2e2d8e580a7843065   
140524  zawfotcx           0a67144ee20745181efe19f9e3b9ac0929b578d4   
140525  x51jorbm           047ca5a826da314da2421d2101ae6cef6ec6644d   
140526  1f96hzfz           9057e

<a id='Single_valued_and_unique'></a>
### Single valued and unique

First start by making sure the columns that are supposed to be single valued and unique are actually so.

In [5]:
SingleValued_and_Unique = ['cord_uid', 'pmcid', 'pubmed_id', 'mag_id', 'who_covidence_id', 'arxiv_id', 'url', 's2_id']

In [6]:
for Column in SingleValued_and_Unique:
    print(Column, metadataFrame[Column].dropna().is_unique)

cord_uid False
pmcid True
pubmed_id False
mag_id True
who_covidence_id True
arxiv_id True
url False
s2_id False


All except mag_id, who_covidence_id and arxiv_id appear to be non-unique. Less than optimal...

Start by checking the cord_uid.

In [7]:
ValueCounts = metadataFrame['cord_uid'].value_counts()
ValueCounts[ValueCounts>1]

uym826bh    6
hgpgeel6    5
lhj7cd8t    5
30sjxtnk    5
3l4mntw2    4
q66wkdkd    4
7plpn6sn    4
4kgndu36    4
453q288g    4
lmo0ry4k    4
ftanlk8k    4
hkm8yspk    4
opwd03wp    4
9naohwpo    4
4wsueam1    4
tnctqc1a    4
q6px4r3j    4
v88f9e7m    4
hl8v7n66    4
mejio1m7    4
c9jg3bf3    4
kriro0t4    4
w0brwyby    4
ciqrsfpo    4
f8hgcngj    4
8xz0ddci    4
trzkb9ne    4
ux7bdk9x    4
3msk5ggc    4
klseuylw    4
           ..
05xedq1l    2
ziepfnpz    2
ucorpfw5    2
6donibpz    2
nn2an4up    2
ydq687u6    2
q5dhugfa    2
p370gjjn    2
28ygsbo1    2
3xe2jjl4    2
o6bkxn7l    2
ojtp90na    2
2tu707ng    2
eich19nx    2
7xb7hj9u    2
pauzgepm    2
bim76jna    2
h10o18ss    2
olveh2gf    2
ywkavncn    2
56lr1rlf    2
l0ggbztj    2
cb56np26    2
0irr4tnw    2
a6o88zlr    2
qqh7ayga    2
l0ongbag    2
0n5bxp93    2
ft9hclnz    2
fcd07krs    2
Name: cord_uid, Length: 781, dtype: int64

As this is supposed to the persistent identifier, looks like some duplicates happening. Close to 800 in fact.

So the easiest way to start with this is by checking to see if these are duplicates, and if they are just eliminating the double records.

In [8]:
metadataFrame[metadataFrame['cord_uid'].duplicated(keep=False)].sort_values(by=['cord_uid'])

cord_uid                                                sha  \
27100   028avudf                                                NaN   
60114   028avudf                                                NaN   
134087  028avudf           4e11d7fbd071bf098dddb9e5beca620fcdaacc39   
97788   04jbodmf                                                NaN   
65413   04jbodmf                                                NaN   
105132  052xdp69           9d9d15ee5ecd1555d091ef0a2a36485d8c810e49   
61380   052xdp69           9d9d15ee5ecd1555d091ef0a2a36485d8c810e49   
61379   052xdp69                                                NaN   
51971   05my504t                                                NaN   
56347   05my504t                                                NaN   
65337   05p0hgvq                                                NaN   
59683   05p0hgvq                                                NaN   
127725  05xedq1l           17de15d04f7a7643550666d2f0da773f5b7dc95c   
65404   05xedq1l                                                NaN   
65772   06vqlhoz                                                NaN   
116298  06vqlhoz                                                NaN   
65940   07l9hqsr                                                NaN   
131408  07l9hqsr           7eb8da93320ea097a81497d97ed22bdad6e708dc   
108349  08hvlnff           9d6692c99e291964c6eac6a314e3b602ea00acf9   
38353   08hvlnff                                                NaN   
66472   08o0re45           57187437fad015cfa48ce55168ac6e60c9f687a4   
106728  08o0re45           57187437fad015cfa48ce55168ac6e60c9f687a4   
63377   097nifny                                                NaN   
137296  097nifny           308576d6375abb65564b6dce15c7e3d5d014cb35   
139546  0cq5ee1i                                                NaN   
64269   0cq5ee1i           6936e1a9ece051a13678f9386d399edea38170ff   
64268   0cq5ee1i                                                NaN   
64404   0dxl6t4p           000eec3f1e93c3792454ac59415c928ce3a6b4ad   
115247  0dxl6t4p           000eec3f1e93c3792454ac59415c928ce3a6b4ad   
30090   0fbmelx0                                                NaN   
...          ...                                                ...   
47226   zhcghyw8                                                NaN   
41140   zhcghyw8                                                NaN   
62532   ziepfnpz           f7f99b10a8f56b2e0e85885a30ef3e3f3258ce7d   
132548  ziepfnpz                                                NaN   
62392   ziijht2j                                                NaN   
130094  ziijht2j           0ad7c9ac2c3ced6c5cee4267796228ca532da097   
106317  zmokxa10           831875bf3c05bcfb942f2a35f4b066a1e4cb87d7   
60379   zmokxa10                                                NaN   
60380   zmokxa10           831875bf3c05bcfb942f2a35f4b066a1e4cb87d7   
65087   zn87f1lk                                                NaN   
140441  zn87f1lk           1cf54d1c77b7f0494ab971429d26e0e199952d09   
40643   zpib1us1                                                NaN   
112901  zpib1us1           9ec51cd6ce837abb8eb9bede87b12bf6ea05ed4d   
60684   zpqrm0a3                                                NaN   
109019  zpqrm0a3  2d709b295d267319f77e9aa399066b54e34dd265; e87b...   
113179  zq2h0avw  294cdeb31114bf4946dfdb239252a8ffe49cbdf7; dbaf...   
60567   zq2h0avw                                                NaN   
126173  zqbume1i           314260241c5b8b3bf04c0d1f8019893c80c26bda   
61238   zqbume1i           314260241c5b8b3bf04c0d1f8019893c80c26bda   
97254   zrf8lam8                                                NaN   
63144   zrf8lam8                                                NaN   
63145   zrf8lam8                                                NaN   
56991   zssgqpv7                                                NaN   
4422    zssgqpv7           c410269beb98d142c4248faf1e0e212df05c585d   
33971   zwkl1ywk                

It looks like this is primarily an issue of source.

For now just drop these records. A little under 800 will not have an impact within a corpus of 140k+

In [9]:
print('Rows before: {}'.format(metadataFrame.shape[0]))
metadataFrame = metadataFrame.drop_duplicates('cord_uid', keep=False) # drop all duplicates. keep='first' or keep='last' would keep the first/last
print('Rows after: {}'.format(metadataFrame.shape[0]))

Rows before: 140532
Rows after: 138751


In [10]:
metadataFrame['cord_uid'].dropna().is_unique

True

So that worked. cord_uid is now unique.

In [11]:
for Column in SingleValued_and_Unique:
    print(Column, metadataFrame[Column].dropna().is_unique)

cord_uid True
pmcid True
pubmed_id False
mag_id True
who_covidence_id True
arxiv_id True
url True
s2_id False


Also cleared up pmcid and url. Would have been nice if it cleared up a few of the others too though. Let's do a quick check of what discrepancies are left.

In [12]:
# For each column in SingleValued_and_Unique, this prints the values that appear in more than 1 record.
for Column in SingleValued_and_Unique:
    ValueCounts = metadataFrame[Column].value_counts()
    print(ValueCounts[ValueCounts>1])

Series([], Name: cord_uid, dtype: int64)
Series([], Name: pmcid, dtype: int64)
32419926.0    3
16575523.0    2
17098273.0    2
23842446.0    2
1655870.0     2
16115318.0    2
15313178.0    2
8546012.0     2
26499340.0    2
8545971.0     2
11918861.0    2
207058.0      2
16116432.0    2
16112641.0    2
16112593.0    2
1331160.0     2
1656597.0     2
16112259.0    2
9058755.0     2
16111992.0    2
15316497.0    2
26509109.0    2
1655802.0     2
9050994.0     2
26475793.0    2
18095180.0    2
16417929.0    2
2997998.0     2
32238613.0    2
16118891.0    2
             ..
12683406.0    2
25458561.0    2
16964581.0    2
17030494.0    2
16534565.0    2
6171237.0     2
17037549.0    2
2446421.0     2
6199888.0     2
17037602.0    2
16781410.0    2
12484662.0    2
17037504.0    2
17002511.0    2
25558126.0    2
12781012.0    2
196571.0      2
17037499.0    2
17037567.0    2
17922397.0    2
16963099.0    2
16582497.0    2
16781449.0    2
12377851.0    2
17052935.0    2
16824137.0    2
16940861.

__pubmed_id__
This has almost 7000 duplicates. Lets have a quick look.

In [13]:
# These get a bit complicated. Basically it starts by selecting the rows that have duplicate (non-null) pubmed_id

metadataFrame[metadataFrame['pubmed_id'].duplicated(keep=False) & metadataFrame['pubmed_id'].notnull()].sort_values(by=['pubmed_id'])

cord_uid                                       sha           source_x  \
74183   crz30eba                                       NaN            Medline   
137250  nl5kmmtd  3a2ff3a86c4b06fb35f0422034ce8c65f7ca0bd3      Elsevier; PMC   
76159   4l35bv99                                       NaN            Medline   
136565  m9j3pcj1  5610c53dce08ee3038eade1231d7d6d0fc6a8ad9      Elsevier; PMC   
78077   3e7gpq2z                                       NaN            Medline   
13229   smqmp1mm  8090cf4603341df5350674cd0ab3f9877f67f95e                PMC   
125393  747jppsc  7747dfcb318bbbf7f9edebb69bc91c4b60f5e658      Elsevier; PMC   
79613   u07xr5nt                                       NaN            Medline   
83782   100z5zox                                       NaN            Medline   
97482   h51cebpe  71d3406258973d7ea61a630bf16f5cfa04f78a13      Elsevier; PMC   
78542   d52nz2mf                                       NaN            Medline   
117722  v8nifzwn  937e8370900dc469c498f186ae7ede00332e6584      Elsevier; PMC   
81708   3jfdx6w0                                       NaN            Medline   
9715    tc402gn4  f7d6f18903a47360a5faff11192ba23e203e5fdf                PMC   
6196    fe24ezt6  e9854fb393df6ec24710f80621ac88983c3a0fcc                PMC   
77977   aq698q9r                                       NaN            Medline   
75395   ol9ifnvq                                       NaN            Medline   
97722   j3p8uj0c  3dfa916237b9663c3b6f17604d5e970bc4053898      Elsevier; PMC   
6285    38rgcc5h  34ddcd68407fb141f82a9e14a071c671db883264                PMC   
70834   0s0ebyla                                       NaN            Medline   
6015    d6f6a5ds  76a7b0a301baf4436319df3e013755e6f69a80ca                PMC   
79722   9zr35wfg                                       NaN            Medline   
83799   p6i8srdt                                       NaN            Medline   
128930  z58u1u12  b69e28669ea8392a9fa48373b0c4b9e94335914d      Elsevier; PMC   
119615  4keq0c6u                                       NaN      Elsevier; PMC   
71378   w0nobqu5                                       NaN            Medline   
67916   fnkotkdy                                       NaN            Medline   
130181  6eyetb5i  d62517543cff4abed85182bc70d1b5cfb15e0e6b      Elsevier; PMC   
74705   k4no88d0                                       NaN            Medline   
12188   qdh7okux  da8bd99cf1771969654dc5479f60d95636525c05                PMC   
...          ...                                       ...                ...   
106429  nnftb092                                       NaN  Medline; PMC; WHO   
82776   qmmeae8c                                       NaN            Medline   
67026   zgqbfmzl  08db45ce51ed0b3b32df8fb1b630dfbe2f7a4dff                PMC   
67025   i8lgih81  630b2c2b88928c1c6196c8df498531a636b8019e                PMC   
75377   l6r011by                                       NaN            Medline   
69220   grt4vjck                                       NaN            Medline   
67024   h0boxuse  aab8259cf3330a0e6d48218b13615fc684b5c016                PMC   
72052   9fqo7dj8                                       NaN            Medline   
67027   9alqngxk  8bdb3b4c901d599b20118528573e28c0833dfcf0                PMC   
106029  f6muhf3d  7d5cc36c59956980c1b44c7de01bba774557eab9  Medline; PMC; WHO   
55591   3gzh2m4w                                       NaN            Medline   
81252   62f4tndc                                       NaN            Medline   
13683   l405c850  88b5c94473256dc08e78e5344fd415773224045a                PMC   
82999   wnlsxv7z                                       NaN            Medline   
13681   apw1q940  967f4957bd4dd15955a22f8b644770397e04c407                PMC   
13682   wl4y44l2  77626f41463222b88a4dae970efd16d14e3b5993                PMC   
80887   w2jsxq03                                       NaN            Medline   
139910  t4v368v2  b43c0006285d8b8135a8c

This looks like a case where one has a pmcid and the other does not.

I would like to keep the record that has a document parse.

What I am going to do is a bit tricksy.

I sort the dataframe by the parsefile columns. Then I use the duplicated funciton to keep only the first occuring row for each pubmed_id.

In [14]:
print('Rows before: {}'.format(metadataFrame.shape[0]))
metadataFrame = metadataFrame[~(metadataFrame.sort_values(by=['pdf_json_files', 'pmc_json_files']).duplicated('pubmed_id', keep='last') & metadataFrame['pubmed_id'].notnull())]
print('Rows after: {}'.format(metadataFrame.shape[0]))

Rows before: 138751
Rows after: 131846


It eliminated about 7000 rows, matching what number of duplicates we found above.

In [15]:
metadataFrame['pubmed_id'].dropna().is_unique

True

And, indeed, pubmed_id is now unique.

Now turning to __s2_id__, which recalling from above had over 13k duplicates.

In [16]:
metadataFrame[metadataFrame['s2_id'].duplicated(keep=False) & metadataFrame['s2_id'].notnull()].sort_values(by=['s2_id'])

cord_uid                                                sha  \
74070   mo0bhc4s                                                NaN   
62618   7d7ewkib           adae39fea7d473e279cfffb02a95617e2bd5d65c   
134234  kllv27bl           aa7094234bc71c19d615cc130ca8c2ef98f543a9   
85508   34qmk3jx           c1408f066c54987b04e0b997c6b38197f7ad6e3f   
78579   t92nyi3g                                                NaN   
64327   auasueaw                                                NaN   
85421   lo3m6csl           466dc1e5c227ca49cd0fc535dd402f5049b9161b   
110473  1lojd0xa           a3778f543ef74ff3a0dd90250221b77ee38d7533   
85420   kto778l5           081fce9252d6f843d2329db65ae188d41bf4c30f   
95650   5qp7vk0c           65078532d15de7dbfa1931a5e6e6eba317cee5e3   
82388   pzsezf32                                                NaN   
79148   595rmnjd                                                NaN   
64549   ij8og0zc                                                NaN   
81677   37yutynv                                                NaN   
129358  ef4xg3q0           8167b94a7c3592a7ec805a977913bd34abb9e199   
90870   1vtzahag           f3755a2f5ff6b3027b24baecfec0d1e412e1367b   
120792  nrhq7lqg           1d613eb26d86e376a5f8cb5fdb673b7254110936   
69005   sib9nnar                                                NaN   
60691   k1xxz4hc           2d0bb0da2503469a9aa8e94a264c0f37d100ccba   
92932   34vikc2n           5d7dc504177db9326327e5cc7417bfa045e0f8ad   
133246  u4z9ijrr           eb84579d04e12484426a03cb27a4799d2c649f04   
74611   jl2ro0jn                                                NaN   
74608   31qldfs1                                                NaN   
74609   9bb88qla                                                NaN   
74610   77gibaow                                                NaN   
94591   cfvw4crl           a4de3a6fc3e5ef424dd225e58b16b3c8049d2db5   
110422  s3t2kdww  dc1e16996462dd1445f21c419b6e40dd0dd84c02; 354a...   
116136  v9f5jck9                                                NaN   
74633   jsjqfwcz                                                NaN   
61320   q59axo1w           6fded905459e05a94147a83debeef8c4754f5263   
...          ...                                                ...   
131932  keb71cvg                                                NaN   
86134   i058rghi                                                NaN   
133394  rtduw3ly                                                NaN   
85278   uip9sca1                                                NaN   
85279   y5oikl9f                                                NaN   
56481   1b57eqj3                                                NaN   
95637   tmu9ygc5                                                NaN   
87175   ttiz1gs4                                                NaN   
87174   d8w53x4e                                                NaN   
114921  ns957xmv                                                NaN   
93050   q2ctl2cu                                                NaN   
96406   6ld64bbb                                                NaN   
96407   e68xqpxt                                                NaN   
93109   bw8v7dr9                                                NaN   
93108   qplgmh3y                                                NaN   
95736   25w2ap1k                                                NaN   
95737   qm3wqtab                                                NaN   
95738   rrbxoylv                                                NaN   
85316   9887fsv2                                                NaN   
85317   gnu2ljp8                                                NaN   
93082   yb313vfh                                                NaN   
93081   mhx9bao2                                                NaN   
85307   t7q3scsi                                                NaN   
85306   01pghohv                                                NaN   
87607   gpv5zjfj                

The discrepancies here seem to arise from cases where Semantic Scholar has (seemingly correctly) grouped a published paper with its ArXiv version but the abstract differed between the two versions. This is really some tricky territory. My inclination at the moment is to leave this as is, and circle back to it if necessary.

Lets finish by just checking again all the columns that should be single valued and unique.

In [17]:
for Column in SingleValued_and_Unique:
    print(Column, metadataFrame[Column].dropna().is_unique)

cord_uid True
pmcid True
pubmed_id True
mag_id True
who_covidence_id True
arxiv_id True
url True
s2_id False


So other than s2_id all are now unique.

In [18]:
metadataFrame.shape[0]

131846

And have arrived at 131,846 records.

<a id='Single_valued'></a>
### Single valued

Based on investigations to the this point, it looks like they are denoting multiple values within a column via a semicolon (;). Thus it is likely best to quickly check to make sure the columns we inferred to be inferred to be single valued are actually so.

In [19]:
SingleValued = ['title', 'doi', 'license', 'publish_time']

In [20]:
for Column in SingleValued:
    print(Column, metadataFrame[Column].str.contains(';').fillna(False).sum())

title 575
doi 25
license 0
publish_time 0


license and publish_time are single valued.

doi has a few cases that are not, will have a look but likely ok to just leave as a single valued variable and deal with the corner case if we end up working with doi.

title need to be investigated.

In [21]:
metadataFrame[metadataFrame['doi'].str.contains(';').fillna(False)]

cord_uid                                       sha       source_x  \
8394    no7mnxtj  a1bf6ada1b47e286db751f32fb6d74a6a41328b4            PMC   
12259   djxoiytq  64066db1148876bb85f1c65e2acfab11dfd89c11            PMC   
12262   lreguu45  01207321329715e944d0ae2cba853b919340ad91            PMC   
12422   23cdi61w  11c686b3de7f984cb5e82860014a23050d7cf740            PMC   
12435   2cv7cs48  a0e60329e1527b513e6bfcc6f5da46c7bbf19bf8            PMC   
12444   cuy6dihq  279b4d3afbe2142f1ca72d7536b78b042ce35529            PMC   
12451   khxb15wp  c5bcf6c3957245b4a27e4cadb72c7f1181fc1a32            PMC   
12456   0mw8x66n  e374c4e801f0bd2473b833d5aa9763fa9a1fcec0            PMC   
12563   2rdtj01s  dc0c47c12c9fddc79adc3c14760cae6dd914eaa2            PMC   
13103   d5eo1jv7  bd0283a74bd15c8b481ce8dc2d72d699cdcc7c86            PMC   
13106   eeq50vi0  c06d625e79097828798cc8b814c9ab9f2581b1e8            PMC   
13388   7nusffwg  660c54e43898f9b7c472f90d2858bcb19b2768bb            PMC   
13393   xxwc9ysp  6a8accf080085c9b9d421dfef3fbfd4d51e660dd            PMC   
13720   v862bw1o                                       NaN            PMC   
16286   lzj56lm6                                       NaN            PMC   
16287   1gf1l0to                                       NaN            PMC   
19979   2v01j64f                                       NaN            PMC   
23645   vgfv5pi4  5b28b06755683f53adeb5b5181903d5e45f577fb            PMC   
23699   orjzvvau  4a7d736768e4bd25568077815a7254dad2b9e6a1            PMC   
24066   d5pl7ie6  21c939104a13c465e95875671a281bb4d6506caa            PMC   
53317   gcarnb6w                                       NaN        Medline   
103791  2f5w411y  a9c2a7e2afbc82f6e729fd078eeb336946227116  Elsevier; PMC   
110163  bropr0dy  2f185371b60849f8846c0b57934ae863ea7274ba   Medline; PMC   
111937  ybhcwln6                                       NaN   Medline; PMC   
125331  gihp7ulm  042a57ce54d353b67f9faff39f7829c45dc7b101   Medline; PMC   

                                                    title  \
8394    Generation of an immortal differentiated lung ...   
12259   Molecular mimicry and multiple sclerosis: Dege...   
12262   Increased serum levels of cartilage oligomeric...   
12422   Electron microscopic observations on structure...   
12435   Tubular complexes of endoplasmic reticulum in ...   
12444   Kaposi's sarcoma in renal transplant recipient...   
12451   Morphologic study of virus‐like particles in a...   
12456                Introduction to virus‐caused cancers   
12563   The role of ecological theory and practice in ...   
13103   A Review of the Magnetic Relaxation and Its Ap...   
13106   Factors Influencing the Size Structure of Broo...   
13388   Classification of Texas Reservoirs in Relation...   
13393   Seasonal Diversity, Abundance, and Distributio...   
13720   Pediatric Patient Safety in the Ambulatory Set...   
16286          Invasive Species: The Search for Solutions   
16287   Global Zoonotic Disease Surveillance: An Emerg...   
19979                 Atlas of Travel Medicine and Health   
23645   SIMULATING NEW ZEALAND FOREST DYNAMICS WITH A ...   
23699   Conservation Genetics and the Management of En...   
24066   Water regimes and littoral plants in four weir...   
53317   Polyvalent Interactions in Biological Systems:...   
103791  Exotic Viral Diseases: A Global Guide Stephen ...   
110163  Guidelines for managing lesser prairie‐chicken...   
111937                                         Dispatches   
125331  HIGH-FLOW NASAL CANNULA POST-TRACHEAL EXTUBATI...   

                                                      doi       pmcid  \
8394      10.1290/1071-2690(2000)036<0374:goaidl>2.0.co;2  PMC7101677   
12259   10.1002/1531-8249(199905)45:5<559::aid-ana3>3....  PMC7159663   
12262   10.1002/1529-0131(199803)41:3<544::aid-art21>3...  PMC7159671   
12422   10.1002/1097-0142(197106)27:6<1449::aid-cncr28...  PMC7162171   
12435   10.1002/1097-0142(19820415)49:8<1629::aid-cnc

So the __doi__ cases containing ';' just look like mis-parsings. Will assume doi is single valued and come back to this matter if we end up having to use doi for something.

Turning to __title__:

In [22]:
metadataFrame[metadataFrame['title'].str.contains(';').fillna(False)]

cord_uid                                                sha  \
1840    fkqdrwer           5e0504a1c581c36bbf2c2c97439b87b9834b108b   
2271    s4qki9l5           f4c43e4ae49ca69dbac32620bd0a73ecbb683b91   
2509    l2dc9x12           d760a0c49650a6c2a3d822bab4417057e6be1cd1   
2615    9gftenfz  f4ac2b1e9b9523671d7c38c27a19f1eec2e15d75; 6044...   
2880    k81qwzof           a7154cf3907686df4aab059a0be568e0dacb44cb   
3515    n7ev5osc           7018967c7ebc76e4c6d3decf7e836957ff427602   
3641    muozucj7                                                NaN   
4377    eoz1l5fi           041e50456e5c62980ff8bb61d83283f58f09ef6d   
4756    q10zmids           3061f05203159384dfbb2fd9b1d9a1ca7b98c8a6   
5456    5fk3clav                                                NaN   
5509    l1hmki8o                                                NaN   
5612    i696fyja                                                NaN   
5656    p4s5uqz8           6cc29c53c7f0a16b790c0bc685b6ca05abd98d7a   
5931    ptsuhbod           7a127771726a4021821b1741661518a10cc1f73f   
7909    5pzcfcoe           541f2811c88ee0894e8ef20190edcf403125f902   
9994    187o4ukz                                                NaN   
10416   r1haxog5                                                NaN   
10603   hjkx4k1a           d017f18e6abdd511226d2fddfd8b797499484fc4   
10729   cweeatnh           f275fa315d349056864929245493ee422a2f5c4d   
11670   3l62b03x                                                NaN   
11930   8v25etag           00fddd1ce0dae85355b9fdf8bbbe6bed4475f7e8   
11984   ub4p4ngr                                                NaN   
12311   s2db8r85                                                NaN   
12359   n4wqy6vh                                                NaN   
12605   xq3951b3                                                NaN   
12625   rgzccqp3           7f2863b59185c97f0e948a9caa32c0deb3e1a5a0   
12713   cla2ta1b                                                NaN   
13021   qcfts6fu           1038b97923cff88c849b94467976ba1f93a3ceda   
13286   1cxrnrv3           122ed2c3ce7840bf533c05931efe085f2e74d7d3   
13515   vhqc5svz           2a1af58bd788fca168913f1d5de3a0a0f707c735   
...          ...                                                ...   
130195  eci120hv                                                NaN   
130401  sbsyzbse                                                NaN   
130588  jb455t9p           05513da119ba8f48c1c468a8a40a591b731c4479   
131072  1yarrsep           a3bbdeb8416af69d5dd2a752af1c87eb195c5fa4   
131089  2bzh4tqg           2182aebd9ddd3f56f1898087c02430e1e92cf205   
131119  lxtqc77m                                                NaN   
131361  9ibswt32           915bbf3af94c8bb413b496b0d4d7f88a1ed7d7bb   
131479  uqrewfd5                                                NaN   
131498  knc0ruou           a4660be6df902c34e11658e30722bff0b504429a   
132429  vx8z54m7                                                NaN   
132523  avov4yxv           322c0f442764f80b0ba7408dff7e0741e0246d6d   
132554  6o9grg76                                                NaN   
132677  9k52e5i5                                                NaN   
133152  kwy9puxo           8b49451747e38074c425e289a7105ac6c4ef297c   
133667  52kq2u8a           8315fb05f84a283b0bdab306586803bd4c6a2c7e   
134029  n0x192pm                                                NaN   
134165  e171z7fh                                                NaN   
134369  84ci8h6p           73f37594450d368d25b6b3c4df0013805d345d73   
134432  0ddutmdd           27bafe2069fadba9638ec9eb5ae1afaf3c78cfb8   
135228  n32nx0bz                                                NaN   
135325  pdqkfhtq  1b8770691fe974dfb269a16db59750370b9c0c21; 1c4e...   
135726  m5h2w6nm                                                NaN   
135827  v6a41kh8           ff4fffe02138b0b232334d997965d11fd936916b   
137714  zedd7ug4           cafe63b787c472f52f88daab2977460e6283bd58   
137908  1bmpvmdd           0c8fe

Lets make these more readable.

In [23]:
for record in metadataFrame[metadataFrame['title'].str.contains(';').fillna(False)]['title']:
    pp(record)

('Induction of Interferon-Stimulated Genes on the IL-4 Response Axis by '
 'Epstein-Barr Virus Infected Human B Cells; Relevance to Cellular '
 'Transformation')
('Exploring the Innate Immunological Response of an Alternative Nonhuman '
 'Primate Model of Infectious Disease; the Common Marmoset')
('Correction: Xie, H.; et al. 3D QSAR Studies, Pharmacophore Modeling and '
 'Virtual Screening on a Series of Steroidal Aromatase Inhibitors. Int. J. '
 'Mol. Sci. 2014, 15, 20927–20947')
('Meta-genomic analysis of toilet waste from long distance flights; a step '
 'towards global surveillance of infectious diseases and antimicrobial '
 'resistance')
('Surfactant Protein C-associated interstitial lung disease; three different '
 'phenotypes of the same SFTPC mutation')
('Correlation of central venous pressure with venous blood gas analysis '
 'parameters; a diagnostic study')
'Authors’ correction for Euro Surveill. 2017;22(11)'
('Award Winners and Abstracts of the 32nd Annual Symposium of The

It looks like most of these are cases where a semicolon (';') is in the place of a colon (':').

If course generalizing from checking a small number can be dangerous, but for now just replace ';' with ':'.

In [24]:
metadataFrame['title'] = metadataFrame['title'].str.replace(';',':')

Just to check them all again:

In [25]:
for Column in SingleValued:
    print(Column, metadataFrame[Column].str.contains(';').fillna(False).sum())

title 0
doi 25
license 0
publish_time 0


This is something we can live with.

<a id='Setting_up_a_schema'></a>
# Setting up a schema

Recalling the fields covered above:

__single valued__, __unique__
* 'cord_uid'
* 'pmcid'
* 'pubmed_id'
* 'mag_id'
* 'who_covidence_id'
* 'arxiv_id'
* 'url'

__single valued__
* 'source_x'
* 'title'
* 'doi'
* 'license'
* 'abstract'
* 'publish_time'
* 'journal'
* 's2_id'

__multi-valued__, __unique__
* 'pdf_json_files'
* 'pmc_json_files'

__multi-valued__
* 'sha'
* 'authors'

The strategy here is to create a main table that stores all the single valued fields, and then build four more that cover the multi-valued fields.

cord_uid will be the primary index across all.

In [26]:
import sqlite3 # For now just going to populate in sqlite. Will probably put down to disk better later in MySQL.

In [27]:
conn = sqlite3.connect('CORD_metadata.db')
cur = conn.cursor()

<a id='Main_table'></a>
## Main table

In [28]:
# Organize the data a bit in a dict
columnData = {}
columnData['cord_uid'] = {'SQLName': 'cord_uid', 'SQLType': 'char(8) NOT NULL UNIQUE', 'PythonType': 'str'}
columnData['pmcid'] = {'SQLName': 'pmcid', 'SQLType': 'char(9) UNIQUE', 'PythonType': 'str'}
columnData['pubmed_id'] = {'SQLName': 'pubmed_id', 'SQLType': 'number(8,0) UNIQUE', 'PythonType': 'int'}
columnData['mag_id'] = {'SQLName': 'mag_id', 'SQLType': 'number(10,0) UNIQUE', 'PythonType': 'int'}
columnData['who_covidence_id'] = {'SQLName': 'who_covidence_id', 'SQLType': 'char(12) UNIQUE', 'PythonType': 'str'}
columnData['arxiv_id'] = {'SQLName': 'arxiv_id', 'SQLType': 'text UNIQUE', 'PythonType': 'str'}
columnData['url'] = {'SQLName': 'url', 'SQLType': 'text UNIQUE', 'PythonType': 'str'}
columnData['source_x'] = {'SQLName': 'source_x', 'SQLType': 'text', 'PythonType': 'str'}
columnData['title'] = {'SQLName': 'title', 'SQLType': 'text', 'PythonType': 'str'}
columnData['doi'] = {'SQLName': 'doi', 'SQLType': 'text', 'PythonType': 'str'}
columnData['license'] = {'SQLName': 'license', 'SQLType': 'text', 'PythonType': 'str'}
columnData['abstract'] = {'SQLName': 'abstract', 'SQLType': 'text', 'PythonType': 'str'}
columnData['publish_time'] = {'SQLName': 'publish_time', 'SQLType': 'DATE', 'PythonType': 'str'}
columnData['journal'] = {'SQLName': 'journal', 'SQLType': 'text', 'PythonType': 'str'}
columnData['s2_id'] = {'SQLName': 's2_id', 'SQLType': 'number(9,0)', 'PythonType': 'int'}

The SQL types were arrived at by inspection. It is likely that some can be improved but will circle back as necessary.

In [29]:
TableName = 'main'
cur.execute('DROP TABLE IF EXISTS {}'.format(TableName)) # always start fresh.
CreateTableQuery = 'CREATE TABLE {}('.format(TableName)
for Column in metadataFrame.columns:
    if Column in columnData:
        CreateTableQuery += '{SQLName} {SQLType}, '.format(**columnData[Column])
CreateTableQuery = CreateTableQuery[:-2]+')' # hate this [:-2] construction to eliminate the last ','. Could be impoved with a join on a list.
#print(CreateTableQuery)
cur.execute(CreateTableQuery)

In [30]:
# This is work around for correctly inserting NULL.
def FormatNULL(Value, pythonType):
    if pd.isnull(Value):
        return('NULL')
    else:
        if pythonType == 'str':
            return('"{}"'.format(Value.replace('"', "'")))
        else:
            return(str(Value))

In [31]:
# This one can be used for testing. Will exceed IO rate if run.
def INSERT_from_df_Single(_dataframe, _tableName, _columnData, _cursor):
    InsertQueryStub = 'INSERT INTO {} ({}) VALUES '.format(_tableName, ', '.join([col for col in _dataframe.columns if col in _columnData]))
    for Record in _dataframe.to_dict('records'):
        InsertQuery = InsertQueryStub + '({})'.format(', '.join([FormatNULL(Record[col], _columnData[col]['PythonType']) for col in _dataframe.columns if col in _columnData]))
        #print(InsertQuery)
        _cursor.execute(InsertQuery)

In [32]:
# This should be used to actually do the insert.
# Have updated this to work from a list and print when batchsize is hit.
def INSERT_from_df_Mass(_dataframe, _tableName, _columnData, _cursor, _batchSize):
    MassInsertQueryStub = 'INSERT INTO {} ({}) VALUES '.format(_tableName, ', '.join([col for col in _dataframe.columns if col in _columnData]))

    MassInsertValues = []
    for Record in _dataframe.to_dict('records'):
        MassInsertValues.append('({})'.format(', '.join([FormatNULL(Record[col], _columnData[col]['PythonType']) for col in _dataframe.columns if col in _columnData])))

        if len(MassInsertValues) >= _batchSize:
            MassInsertQuery = MassInsertQueryStub + ', '.join(MassInsertValues[:_batchSize])
            _cursor.execute(MassInsertQuery)
            del MassInsertValues[:_batchSize]

    MassInsertQuery = MassInsertQueryStub + ', '.join(MassInsertValues[:_batchSize])
    _cursor.execute(MassInsertQuery)

In [33]:
INSERT_from_df_Mass(metadataFrame, TableName, columnData, cur, 500) # Batches of 500

AttributeError: 'float' object has no attribute 'replace'

In [ ]:
cur.execute('SELECT COUNT(*) FROM {}'.format(TableName))
pp(cur.fetchone())

The right number of records.

In [ ]:
cur.execute('SELECT * FROM {} WHERE publish_time BETWEEN "2001-01-01" AND "2002-01-01"'.format(TableName))
pp(cur.fetchone())

That all lines up quite nicely.

<a id='Multivalued_field_tables'></a>
## Multivalued field tables

<a id='pdf_json_files'></a>
### pdf_json_files

In [ ]:
columnData = {}
columnData['cord_uid'] = {'SQLName': 'cord_uid', 'SQLType': 'char(8) NOT NULL', 'PythonType': 'str'}
columnData['pdf_json_files'] = {'SQLName': 'pdf_json_files', 'SQLType': 'text', 'PythonType': 'str'}

In [ ]:
MainIndex = 'cord_uid'
MultiValuedVariable = 'pdf_json_files'
TableName = MultiValuedVariable
cur.execute('DROP TABLE IF EXISTS {}'.format(TableName))
CreateTableQuery = 'CREATE TABLE {}({} {}, {} {})'.format(TableName, MainIndex, columnData[MainIndex]['SQLType'], MultiValuedVariable, columnData[MultiValuedVariable]['SQLType'])
#print(CreateTableQuery)
cur.execute(CreateTableQuery)

In [ ]:
# This is work around for correctly inserting NULL.
def FormatNULL(Value, pythonType):
    if pd.isnull(Value):
        return('NULL')
    else:
        if pythonType == 'str':
            return('"{}"'.format(Value.replace('"', "'")))
        else:
            return(str(Value))

In [ ]:
def INSERT_from_df_MultiValue_Mass(_dataframe, _tableName, _mainIndex, _multiValueColumn, _columnData, _cursor, _batchSize):
    MassInsertQueryStub = 'INSERT INTO {} ({}) VALUES '.format(_tableName, ', '.join([_mainIndex, _multiValueColumn]))

    MassInsertValues = []
    for Record in _dataframe.to_dict('records'):

        if pd.isnull(Record[_multiValueColumn]):
            MassInsertValues.append('({})'.format(', '.join([FormatNULL(Record[col], _columnData[col]['PythonType']) for col in [_mainIndex, _multiValueColumn]])))
        else:
            for Value in Record[_multiValueColumn].split(';'):
                MassInsertValues.append('({}, {})'.format(FormatNULL(Record[_mainIndex], _columnData[_mainIndex]['PythonType']), FormatNULL(Value.strip(), _columnData[_multiValueColumn]['PythonType'])))

        if len(MassInsertValues) >= _batchSize:
            MassInsertQuery = MassInsertQueryStub + ', '.join(MassInsertValues[:_batchSize])
            _cursor.execute(MassInsertQuery)
            del MassInsertValues[:_batchSize]

    MassInsertQuery = MassInsertQueryStub + ', '.join(MassInsertValues[:_batchSize])
    _cursor.execute(MassInsertQuery)

In [ ]:
INSERT_from_df_MultiValue_Mass(metadataFrame, TableName, MainIndex, MultiValuedVariable, columnData, cur, 500)

<a id='pmc_json_files'></a>
### pmc_json_files

In [ ]:
columnData = {}
columnData['cord_uid'] = {'SQLName': 'cord_uid', 'SQLType': 'char(8) NOT NULL', 'PythonType': 'str'}
columnData['pmc_json_files'] = {'SQLName': 'pmc_json_files', 'SQLType': 'text', 'PythonType': 'str'}

In [ ]:
MainIndex = 'cord_uid'
MultiValuedVariable = 'pmc_json_files'
TableName = MultiValuedVariable
cur.execute('DROP TABLE IF EXISTS {}'.format(TableName))
CreateTableQuery = 'CREATE TABLE {}({} {}, {} {})'.format(TableName, MainIndex, columnData[MainIndex]['SQLType'], MultiValuedVariable, columnData[MultiValuedVariable]['SQLType'])
#print(CreateTableQuery)
cur.execute(CreateTableQuery)

In [ ]:
INSERT_from_df_MultiValue_Mass(metadataFrame, TableName, MainIndex, MultiValuedVariable, columnData, cur, 500)

<a id='sha'></a>
### sha

In [ ]:
columnData = {}
columnData['cord_uid'] = {'SQLName': 'cord_uid', 'SQLType': 'char(8) NOT NULL', 'PythonType': 'str'}
columnData['sha'] = {'SQLName': 'sha', 'SQLType': 'text', 'PythonType': 'str'}

In [ ]:
MainIndex = 'cord_uid'
MultiValuedVariable = 'sha'
TableName = MultiValuedVariable
cur.execute('DROP TABLE IF EXISTS {}'.format(TableName))
CreateTableQuery = 'CREATE TABLE {}({} {}, {} {})'.format(TableName, MainIndex, columnData[MainIndex]['SQLType'], MultiValuedVariable, columnData[MultiValuedVariable]['SQLType'])
#print(CreateTableQuery)
cur.execute(CreateTableQuery)

In [ ]:
INSERT_from_df_MultiValue_Mass(metadataFrame, TableName, MainIndex, MultiValuedVariable, columnData, cur, 500)

<a id='authors'></a>
### authors

In [ ]:
columnData = {}
columnData['cord_uid'] = {'SQLName': 'cord_uid', 'SQLType': 'char(8) NOT NULL', 'PythonType': 'str'}
columnData['authors'] = {'SQLName': 'authors', 'SQLType': 'text', 'PythonType': 'str'}

In [ ]:
MainIndex = 'cord_uid'
MultiValuedVariable = 'authors'
TableName = MultiValuedVariable
cur.execute('DROP TABLE IF EXISTS {}'.format(TableName))
CreateTableQuery = 'CREATE TABLE {}({} {}, {} {})'.format(TableName, MainIndex, columnData[MainIndex]['SQLType'], MultiValuedVariable, columnData[MultiValuedVariable]['SQLType'])
#print(CreateTableQuery)
cur.execute(CreateTableQuery)

In [ ]:
INSERT_from_df_MultiValue_Mass(metadataFrame, TableName, MainIndex, MultiValuedVariable, columnData, cur, 500)

Those are all loaded quite well.

If it ends up necessary to go a step further it would be best to streamline this with a tool like SQLAlchemy.

<a id='Quick_test'></a>
## Quick test

In [ ]:
cur.execute('SELECT * FROM {} WHERE publish_time BETWEEN "2010-01-01" AND "2011-01-01"'.format('main'))
Record = cur.fetchone()
pp(Record)

In [ ]:
MultiValued = ['pdf_json_files', 'pmc_json_files', 'sha', 'authors']

for Table in MultiValued:
    cur.execute('SELECT {} FROM {} WHERE cord_uid IS "{}"'.format(Table, Table, Record[0]))
    pp(cur.fetchall())

So this appears to all be in order.